src/shared/Microsoft.Git.CredentialManager/Interop/Linux/LinuxCredentialStore.cs

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.Git.CredentialManager.Interop.Linux
+{
+    public class LinuxCredentialStore : ICredentialStore
+    {
+        private readonly ISettings _settings;
+        private readonly IGit _git;
+
+        private ICredentialStore _backingStore;
+
+        public LinuxCredentialStore(ISettings settings, IGit git)
+        {
+            EnsureArgument.NotNull(settings, nameof(settings));
+            EnsureArgument.NotNull(git, nameof(git));
+
+            _settings = settings;
+            _git = git;
+        }
+
+        #region ICredentialStore
+
+        public ICredential Get(string key)
+        {
+            EnsureBackingStore();
+            return _backingStore.Get(key);
+        }
+
+        public void AddOrUpdate(string key, ICredential credential)
+        {
+            EnsureBackingStore();
+            _backingStore.AddOrUpdate(key, credential);
+        }
+
+        public bool Remove(string key)
+        {
+            EnsureBackingStore();
+            return _backingStore.Remove(key);
+        }
+
+        #endregion
+
+        private void EnsureBackingStore()
+        {
+            if (_backingStore is null)
+            {
+                // TODO: determine the available backing stores based on the current session
+                // TODO: prompt for the desired backing store
+                // TODO: store the desired backing store to ~/.gitconfig
+                _backingStore = SecretServiceCollection.Open();
+            }
```

Comment:
> This seems a little odd to need to call EnsureBackingStore before forwarding all these calls if the backing store is a singleton.
> ...
> But coming back to this method, why not put this block or method call in the constructor? 

Here I want the `LinuxCredentialStore` to be lazy because we don't know what store we have access to except at runtime, and there are codepaths that don't touch the cred store at all - we don't want to do some distro/cred store support discovery on every invocation of the application.

Plus, in some of the P/Invoke code I've been playing around with we may need to do some initialisation or `dlsym`/`dlopen` on static type constructors, so we really don't want to have code like that running unless we're sure we have all the required native dependencies and shared libs in place.

> The todo here is to figure out what ICredentialStores are available which also makes the LinuxCredentialStore sort of an odd abstraction (IMO).
It seems like high up in the stack we would have a factory or something determine what credential stores are available and instantiate the desired class. Although maybe this is a design pattern I'm unfamiliar with. Thinking aloud.

The `LinuxCredentialStore` is a proxy for the actual 'backing store'. The reasoning for this is:
1. we might have different actual stores depending on the Linux distro
2. we might have different stores available depending on headless/GUI presence
3. we might want/need to read and write to _multiple_ actual backing stores, or do some extra logic to select a backing store.

I felt that making this its own component, being a proxy for some backing stores, was appropriate, rather than push all that also up into a factory that had to deal with selecting between Windows and macOS.

Reply:
Thanks for the explanations! I think the abstraction reasoning is good. Probably in the future we could use `Lazy` proper.